### ISSUE TITLE
cggi: add lower-level CGGI ops

### ISSUE THREAD

**Issue by j2kun** [google/heir#255]
Right now the CGGI dialect contains only boolean gate ops reflecting the high level CGGI gate bootstrapping API. We also want the CGGI dialect to include lower-level operations:

- An intermediate level containing bootstrap ops, such as programmable bootstrap and vanilla bootstrap
- A lower level containing the CGGI-specific implementations of bootstrap, such as external product

Ops that are part of the bootstrap implementation, but are generic—such as LWE modulus switch, LWE key switch, and RLWE->LWE sample extraction—should be in the `lwe` dialect so that they can be reused in other schemes.

**Comment by asraa** [comment]
This would be great as a CGGI to LWE lowering pass. 

cc @FlorentCLMichel this would seem useful to lower level TFHE hardware accelerators' dialects

**Comment by FlorentCLMichel** [comment]
Thanks @asraa for the mention! I think it would indeed be very useful for hardware accelerators. 

I have a naive question: Are the intermediate and lower levels completely separate, or will there be scope to mix and match operations from the two? The reason I ask is:

* On the one hand, from a HW accelerator point of view, it may be best to stop lowering at the highest level supported by the accelerator and let the firmware take things from there. (For instance, it may be more efficient to have a single BlindRotate instruction on hardware that support it than to decompose it into multiple negacyclic rotations and external products, as the former can make it easier to keep the data pipeline full.)

* On the other hand, it would be good to keep as much flexibility as possible for operations that are not part of a higher-level instruction - for instance, depending on the implementation, it may be better to do key-switching and sample extraction before or after BlindRotate. So I think it would be great to, in a sense, stop the lowering at different points for different operations depending on what the target hardware implements or not.

**Comment by j2kun** [comment]
I don't think we need to have the dialect definition force semantics about what is allowed in an IR. That can be the role of the lowering.

I think some accelerators support bootstrap opaquely, and I imagine the lower level ops are mainly to make it easy to decompose lowerings into smaller steps and re-use components across variants of bootstrap. For example, external product is the same across many CGGI bootstrap implementations, but how external product is used differs based on which tricks you pick, e.g., the BMMP17 trick from https://eprint.iacr.org/2017/1114 which trades off a larger key for fewer invocations of external product. And the AutoHoG stuff is similar.

For a backend like TPU, we would then lower these low-level CGGI implementation details to TPU.

**Comment by ai-mannamalai** [comment]
+1
